opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -29,12 +43,41 @@ impl<T: Number<T>> Default for ValueMap<T> {
 
 impl<T: Number<T>> ValueMap<T> {
     fn new() -> Self {
+        let buckets = std::iter::repeat_with(|| Mutex::new(None))
+            .take(BUCKET_COUNT)
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
         ValueMap {
-            values: Mutex::new(HashMap::new()),
+            buckets: Arc::new(buckets),
             has_no_value_attribute_value: AtomicBool::new(false),
             no_attribute_value: T::new_atomic_tracker(),
         }
     }
+
+    // Hash function to determine the bucket
+    fn hash_to_bucket(key: &AttributeSet) -> u8 {
+        #[cfg(not(feature = "use_hashbrown"))]
+        let mut hasher = DefaultHasher::new();
+        #[cfg(feature = "use_hashbrown")]
+        let mut hasher = AHasher::default();
+
+        key.hash(&mut hasher);
+        // Use the 8 least significant bits directly, avoiding the modulus operation.
+        hasher.finish() as u8
```

Comment:
trying to understand more here.. AttributeSet, which is the key here, is calculating hash already and storing it. That uses the default hasher only, so how does it know that this is using a different hashing algorithm?

Reply:
The `hasher` is conditionally initialized either as `AHasher`, or `DefaultHasher`, based on whether `use_hashbrown` feature is enabled or not. So 

```rust
key.hash(&mut hasher);
```

will use the appropriate hashing algorithm based on the hasher object. passed.

Reply:
https://github.com/open-telemetry/opentelemetry-rust/blob/5e67cae02030a98e255d1e9ac1bd25d9bef70963/opentelemetry-sdk/src/attributes/set.rs#L179

it's just returning the hash already calculated using default hash right?

Reply:
> it's just returning the hash already calculated using default hash right?

No.
```rust
impl Hash for AttributeSet {
    fn hash<H: Hasher>(&self, state: &mut H) {
        state.write_u64(self.1)
    }
}
```

Hasher `H` is a generic/template in the above method, so based on the hasher instance used, the hash would be calculated.